### ISSUE TITLE
Update Tests to conform with new design

### ISSUE THREAD

**Issue by emmdim** [eReuse/devicehub-teal#15]

**Comment by JNadeu** [comment]
List of critical features to test:

1. Test Users/Auth/Agents
- Login views, check token
- Create specific user and agents (organization)
-

2. Test Inventory / Device
- Create inventory (with user information)
- Create devices with information (Keyboard, HHD, CPU, Computer, Mobiles, ...)
- 

3. Test Action
- Create different types of actions 
- Snapshot
- Data Wipe Actions
- Benchmarks
-

4. Test Report/Documents
- Export device info (NonComponent) to csv/excel
- Erasure certificate

5. Test Lots
- Create, delete and modify lots
- Add/Del devices to lots
- 

6. Test Tags
- Create and link a tag to device
-

7. Test Rate
- Compute rate v1 (only computers)
- 

8. Test DeliveryNote
- Needs some specific tests?

9. Test Workbench
- To fix wrong snapshots from workbench of specific Computers

**Comment by JNadeu** [comment]
[First pull request](https://github.com/eReuse/devicehub-teal/pull/37)

**Needs help to fix test device and some other tests..
Majority of tests it seems that finish with the same error**. 
For example if you execute [test_device_model](https://github.com/eReuse/devicehub-teal/blob/6fd981829647b93246e06f3cc287f67a72f0ac69/tests/test_device.py#L34)
Info below:

Last application code execution on [owner_id](https://github.com/eReuse/devicehub-teal/blob/6fd981829647b93246e06f3cc287f67a72f0ac69/ereuse_devicehub/resources/device/models.py#L386) line 389 in Computer class models.

    owner_id = db.Column(UUID(as_uuid=True),
                          db.ForeignKey(User.id),
                          nullable=False,
                          **default=lambda: g.user.id)**

It is as if when initializing a device Computer is not assigned the owner_id (User.id) or something similar. This is a very common error in most tests.

Last line error:

_E       sqlalchemy.exc.StatementError: (builtins.AttributeError) '_AppCtxGlobals' object has no attribute 'user' [SQL: 'INSERT INTO computer (id, chassis, ethereum_address, deposit, owner_id, transfer_state, receiver_id, deliverynote_address) VALUES (%(id)s, %(chassis)s, %(ethereum_address)s, %(deposit)s, %(owner_id)s, %(transfer_state)s, %(receiver_id)s, %(deliverynote_address)s)'] [parameters: [{'id': 1, 'chassis': <ComputerChassis.Tower: 'Tower'>, 'ethereum_address': None, 'receiver_id': None, 'deliverynote_address': None}]]_

**Comment by emmdim** [comment]
Well since we introduced the `owner_id` as necessary attribute (`nullable=False`) you logically need to pass an `owner_id` when you create a Device that is a child of `Computer`. 
It makes sense right?

On how to do this you can check other tests, like [here](https://github.com/eReuse/devicehub-teal/blob/testing/tests/test_user.py).

**Comment by fedjo** [comment]
The default value we assign to `Computer` owner comes from the current logged in user. `g.user.id` is the current logged in user. So this function makes sense if it runs within a request or in general an `ApplicationContext` using Flask terminology. For more infor see [here](https://flask.palletsprojects.com/en/1.1.x/api/#flask.g)

So if you try to create a user outside the devicehub, or without a user logged in this call with fail

**Comment by fedjo** [comment]
As I see in `tests/conftest.py` there is the fixture `auth_app_context` so you can user this fixture (an application context with an authenticated user) to run this test ;-)

**Comment by JNadeu** [comment]
Nice feedback! Thank you @emmdim @fedjo 
I push new commit fixing almost all test_device. Right now I'm trying to fix:

**tests/test_device.py::test_sync_execute_register_no_hid_tag_not_linked FAILED**

**Problem:** 
I think it has something to do with some issue related to the ORM and the use of lazy loading. I think I need to investigate more but if you have some tip or recommendation to fix this problem It will be great.

**Error output:**
_sqlalchemy.exc.StatementError: (sqlalchemy.orm.exc.DetachedInstanceError) Instance <User at 0x7f9a9835add8> is not bound to a Session; attribute refresh operation cannot proceed [SQL: 'INSERT INTO computer (id, chassis, ethereum_address, deposit, owner_id, transfer_state, receiver_id, deliverynote_address) VALUES (%(id)s, %(chassis)s, %(ethereum_address)s, %(deposit)s, %(owner_id)s, %(transfer_state)s, %(receiver_id)s, %(deliverynote_address)s)'] [parameters: [{'id': 1, 'chassis': <ComputerChassis.Microtower: 'Microtower'>, 'receiver_id': None, 'deliverynote_address': None, 'ethereum_address': None}]] (Background on this error at: [http://sqlalche.me/e/bhk3](http://sqlalche.me/e/bhk3))_

**tests/test_device.py::test_get_device FAILED
tests/test_device.py::test_get_devices FAILED** 

**Problem:** 
These two tests are very similar and give the same error. I haven't found a solution because if I add _@pytest.mark.usefixtures(conftest.auth_app_context.__name__)_, it still gives an error for the duplicate user:UserClient that we passed as a parameter in the test. If I remove this parameter, I haven't found the way I should perform the GET of /device correctly

### How to run tests
- To run all mvp tests:
`pytest -v -m mvp`
- To run an individual test:
`pytest tests/test_device.py::test_sync_execute_register_no_hid_tag_not_linked`

### TODO
- [ ] Fix mvp test_device.py
- [ ] Fix mvp test_device_find.py
- [ ] Fix test_snapshot.py::test_erase_privacy_standards_endtime_sort
- [ ] Fix test wokbench features (posting snapshots from real devices)
- [ ] Define missing mvp test
- [ ] Implement all missing mvp test
- [ ] ...

**Comment by forste** [comment]
Hey,

I updated the POSTMAN collection for our API, so it should document now all the relevant ressources. IMHO, the API endpoints (or the services they access) should also be covered by the automatic tests. Alternatively, we could also use postman for creating automated API tests. 

You can view the documentation here (public link): https://documenter.getpostman.com/view/9305901/SzzegLHb

You can also check out the postman collection in the postman app (if you are not part of the Usody team yet, let me know and I will send you a link to join the team)

**Comment by forste** [comment]
@JNadeu I think we should also have tests to check whether the user has the right permissions to access the resource (especially for lots, devices, deliverynotes, tags). not sure if this is implemented already?

**Comment by fedjo** [comment]
The problem for 
**tests/test_device.py::test_sync_execute_register_no_hid_tag_not_linked FAILED**

is created because the `User` object is not present in the `db.session`.

When we create a `Computer` device we also assign the `owner_id` field to the current logged_in `user_id`. When we try to save this device to the db (using the `session.flush()`- see sync.py:L189), SqlAlchemy tries to update ,apart from the `Computer` object, also the `User` object the `Computer` is related to.But the `User` db object is not present to the current db session, so that's why we have the exception.

Long story short: A workaround I managed to overcome this issue was by getting the current logged in user and manually adding in the db.session. Eg
```python
from flask import g

user = g.user
db.session.add(user)
```

the the tests passes.
I am not sure if this is a clear solution or just a workaround for that. I will keep you posted.

**Comment by JNadeu** [comment]
Thanks for you feedback. The test_sync_execute_register_no_hid_tag_not_linked is fixed.
[Commit.](https://github.com/eReuse/devicehub-teal/pull/37/commits/54705284a94be2605ba773f135f67d8f33bfe1c9)

**Comment by JNadeu** [comment]
I have created a new branch (PR) to finish fixing tests, and also to review if there are any critical tests we don't have already implemented.
It would be possible to already close the current [PR](https://github.com/eReuse/devicehub-teal/pull/37), so that we can at least run the automatic tests positively.
